### DISCUSSION TITLE
Expected behavior of TraversalControl values returned by GraphQLTypeVisitor

### DISCUSSION THREAD

**Discussion by bclozel** [graphql-java/graphql-java#3386] [Q&A] [answered]
Hey there!

The Spring for GraphQL project received a report that turns out to be an incompatibility issue with the 21.x versions of GraphQL Java. You can see the complete analysis [in this issue comment](https://github.com/spring-projects/spring-graphql/issues/861#issuecomment-1836766080).

At this point I'm not sure about our own `GraphQLTypeVisitor` implementations and I would like to get some clarity about the expected behavior. When such visitors process the schema, this is done via the `TraverserDelegateListVisitor` which delegates the visited nodes to each registered visitor.

Here is my current understanding of the traversal controls:

* `TraversalControl#CONTINUE` should continue visiting the nodes under the visited node
* `TraversalControl#ABORT` should skip all nodes under the current node
* `TraversalControl#QUIT` should stop entirely visiting the schema

From what I understand, such controls are global and a visitor returning `ABORT` will prevent the next visitors from seeing child nodes. Is that right? In the case of Spring for GraphQL's `ConnectionFieldTypeVisitor`, it returns `ABORT` as a way to ignore a part of the graph. Is this the wrong approach?

Thanks!

**Comment by bbakerman** [comment]
@andimarek is the expert on TraverserControl (since he wrote it with someone from Intuit).

My understanding isas you have outlined and looking at `graphql.util.Traverser#traverse(java.util.Collection<? extends T>, graphql.util.TraverserVisitor<? super T>)` it supports that as well.

So pending further explanation from Andi, I would suggest what you have asked here is correct.

**Comment by andimarek** [comment, marked as answer]
@bclozel yes that is the expected behaviour.
